### PR TITLE
Enable passing connection instance as well as connection config for synchronization

### PR DESCRIPTION
**Pull request by SafalPandey**
## Changes

* Add interface for connection instance and config
* Update imports
* Create util to get config from a connection instance
* Create Connection class to wrap both config and instance of a connection
* Refactor migrator to use Connection class for synchronization
* Create connection object as per conn type and remove common util
* Log when connection instance is received and when connection config is received
* Add documentation in README

## Details

This feature allows us to use sync-db's `synchronize` function as an function call within our `ts` scripts. We can use `synchronize` as follows:

```ts
import * as Knex from 'knex';

import { loadConfig } from '@leapfrogtechnology/sync-db/lib/config';
import { synchronize } from '@leapfrogtechnology/sync-db/lib/migrator';

(async () => {
  // Load sync-db.yml
  const config = await loadConfig();

  const db = Knex({
    client: 'mssql',
    connection: {
      host: 'host',
      port: 'dbPort',
      user: 'userName',
      password: 'password',
      database: 'dbName'
    }
  });

  // Rollback and create all db objects using config.
  await synchronize(config, db, { force: false });

  // Perform other db operations
  // ...
})();
```

## Caveat

Setup and Teardown steps aren't always run within a transaction. **You need to specifically pass a trx object to make sure this happens.**

 ```ts
  await db.transaction(async trx => {
    // Rollback and create all db objects using config.
    await synchronize(config, (trx as any), { force: false });

    // Perform other db operations
    // ...
  });
 ```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -163,6 +167,51 @@ Or,
 $ npm run sync-db
 ```
 
+### With scripts
+
+Use sync-db's `synchronize` function within your `ts` scripts. You can use `synchronize` as follows:
+
+```ts
+import { createInstance } from '@leapfrogtechnology/sync-db/lib/util/db';
```
Comment (mesaugat) [inline README.md]: .

File: src/domain/Config.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import ConnectionConfig from './ConnectionConfig';
+
+/**
+ * Interface for sync-db connection file (connections.sync-db.json).
+ */
+interface Config {
```
Comment (mesaugat) [inline src/domain/Config.ts]: Rename to `SyncDbConfig`. There are too many configuration that we are playing with. Let's make it obvious.

File: src/domain/ConnectionConfig.ts [hunk 3]
```diff
@@ -1,10 +1,17 @@
...
+  id?: string;
+  host: string;
+  port: string;
+  user: string;
+  database: string;
+  password: string;
+  client: string;
+  requestTimeout?: number;
+  charset?: string;
+  options?: any;
```
Comment (mesaugat) [inline src/domain/ConnectionConfig.ts]: It should not be `any`? It should be something like `Knex.options` from Knex types?

File: src/migrator.ts [hunk 4]
```diff
@@ -56,20 +57,20 @@ export function setup(config: Configuration): (connectionConfig: Connection) =>
...
+export function teardown(config: Configuration): (connection: Connection) => Promise<void> {
   const { basePath, sql } = config;
 
-  return async (connectionConfig: Connection) => {
-    const logDb = dbLogger(connectionConfig);
+  return async (connection: Connection) => {
+    const logDb = dbLogger(connection.config);
 
-    logDb(`Running rollback on a database: ${connectionConfig.id}`);
+    logDb(`Running rollback on a database: ${connection.config.id}`);
```
Comment (mesaugat) [inline src/migrator.ts]: Remove `a`

File: src/sqlRunner.ts [hunk 5]
```diff
@@ -7,7 +7,7 @@ import { dbLogger } from './logger';
 import Mapping from './domain/Mapping';
 import SqlCode from './domain/SqlCode';
 import * as promise from './util/promise';
-import Connection from './domain/Connection';
+import ConnectionConfig from './domain/ConnectionConfig';
```
Comment (mesaugat) [inline src/sqlRunner.ts]: .

File: README.md [hunk 6]
```diff
@@ -141,7 +145,7 @@ For local installation you can trigger it with [`npx`](https://www.npmjs.com/pac
 $ npx sync-db
 ```
 
-### Using npm script
+#### Using npm script
```
Comment (mesaugat) [inline README.md]: Revert?

File: src/domain/ConnectionConfig.ts [hunk 7]
```diff
@@ -1,10 +1,19 @@
...
-}
+type KnexConnections = Knex.ConnectionConfig
+  & Knex.MariaSqlConnectionConfig
+  & Knex.MySqlConnectionConfig
+  & Knex.MsSqlConnectionConfig
+  & Knex.SocketConnectionConfig
+  & Knex.Sqlite3ConnectionConfig;
+
+type ConnectionConfig = KnexConnections
+  & {
```
Comment (mesaugat) [inline src/domain/ConnectionConfig.ts]: Should we make a separate interface?

File: src/util/db.ts [hunk 8]
```diff
@@ -1,17 +1,42 @@
...
-export function createInstance(connectionConfig: Connection): Knex {
+export function createInstance(connectionConfig: ConnectionConfig): ConnectionInstance {
   return Knex({
     client: connectionConfig.client,
     connection: connectionConfig
   });
 }
+
+/**
+ * Reteives connection configurations from a connection instance.
```
Comment (mesaugat) [inline src/util/db.ts]: `Retrieves`

File: src/util/db.ts [hunk 9]
```diff
@@ -1,17 +1,42 @@
...
+  };
+}
+
+/**
+ * Returns true if the provided object is a connection instance.
+ *
+ * @param {any} obj
+ * @returns {boolean}
+ */
+export function isConnectionInstance(obj: any): obj is ConnectionInstance {
```
Comment (mesaugat) [inline src/util/db.ts]: TypeGuard 🎉

File: src/Connection.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
+import { getConfig, createInstance } from './util/db';
```
Comment (kabirbaidhya) [inline src/Connection.ts]: Remove `db.ts` altogether these functions could be made local methods to this class.

File: README.md [hunk 11]
```diff
@@ -163,6 +167,50 @@ Or,
 $ npm run sync-db
 ```
 
+### With scripts
```
Comment (kabirbaidhya) [inline README.md]: Programmatic Usage

File: src/domain/ConnectionInstance.ts [hunk 12]
```diff
@@ -0,0 +1,8 @@
+import * as Knex from 'knex';
+
+/**
+ * Database connectionInstance configuration.
+ */
+type ConnectionInstance = Knex
+
+export default ConnectionInstance;
```
Comment (kabirbaidhya) [inline src/domain/ConnectionInstance.ts]: Delete this.

File: src/util/db.ts [hunk 13]
```diff
@@ -1,17 +1,42 @@
...
   });
 }
+
+/**
+ * Reteives connection configurations from a connection instance.
+ *
+ * @param {ConnectionInstance} db
+ * @returns {ConnectionConfig}
+ */
+export function getConfig(db: ConnectionInstance): ConnectionConfig {
```
Comment (kabirbaidhya) [inline src/util/db.ts]: Move all these to the `Connection` class.

File: src/sqlRunner.ts [hunk 14]
```diff
@@ -102,16 +101,18 @@ export function getDropStatement(type: string, fqon: string): string {
...
+export function runSequentially(
+  files: SqlCode[],
+  connection: Connection
+): Promise<void> {
+  const logDb = dbLogger(connection);
   const promises = files.map(file => {
     logDb(`Running ${file.name}`);
 
-    return trx.raw(file.sql);
+    return connection.instance.raw(file.sql);
```
Comment (kabirbaidhya) [inline src/sqlRunner.ts]:
```
connection.query(file.sql)
```

**Pull request merged**